Update Deploy work pool guide to include `.deploy` (#11174)
PrefectHQ/prefect: docs/guides/prefect-deploy.md

search:
  boost: 2
---

# Deploying Flows to Work Pools and Workers

> In this guide, we will dive into configuring a deployment that uses a work pool for for dynamic infrastructure.

Reviewer: ```suggestion
In this guide, we will configure a deployment that uses a work pool for for dynamically provisioned infrastructure.
```

    A -.-> B
    E -.-> G

    classDef gold fill:goldenrod,stroke:goldenrod,stroke-width:4px,color:black
    classDef yellow fill:gold,stroke:gold,stroke-width:4px,color:black
    classDef gray fill:lightgray,stroke:lightgray,stroke-width:4px
    classDef blue fill:blue,stroke:blue,stroke-width:4px,color:white
    classDef green fill:green,stroke:green,stroke-width:4px,color:white
    classDef red fill:red,stroke:red,stroke-width:4px,color:white
    classDef dkgray fill:darkgray,stroke:darkgray,stroke-width:4px,color:white
```

The work pool types above require a worker to be running on your infrastructure to poll a work pool for scheduled flow runs.

!!!note "Additional work pool options available with Prefect Cloud"

    Prefect Cloud offers other flavors of work pools that don't require a worker:

    - [Push Work Pools](/guides/deployment/push-work-pools) - serverless cloud options that don't require a worker because Prefect Cloud submits them to your cloud provider on your behalf.
    
    <!-- - [Managed Execution]() TK - Prefect Cloud submits and runs your deployment on serverless infrastructure TK - Jake PR in progress -->

> In this guide we focus on deployments that require a worker.

Reviewer: ```suggestion
In this guide, we focus on deployments that require a worker.
```

When creating a deployment that uses a work pool, we must answer _two_ basic questions:

- What instructions does a [worker](/concepts/work-pools/) need to set up an execution environment for our flow? For example, a flow may have Python package requirements, unique Kubernetes settings, or Docker networking configuration.
- How should the flow code be accessed?

## Creating work pool-based deployments

The [tutorial](/tutorial/deployments/) shows how you can create a deployment with a long-running process using `.serve` and how to move to a [work-pool-based deployment](/tutorial/workers/) setup with `.deploy`.
See the discussion of when you might want to move to work-pool-based deployments [there](/tutorial/workers/#why-workers-and-work-pools).

In this guide, we will explore how to use `.deploy` in more depth and discuss `prefect.yaml`, a YAML-based alternative for managing deployments.

Use the tabs below to explore both deployment creation options.

> === ".deploy"

Author: Yeah, good question. I noticed that and am pretty sure there isn't a way to do so. Will double check.

> === ".deploy"

Author: Yeah, nothing doing. Agree it's suboptimal. My sense based on very little evidence is that the TOC doesn't get tons of use. Could remove headings and use html to change text sizes, but that seems worse than the current state.

    ### Automatically bake your code into a Docker image 

    You can create a deployment from Python code by calling the `.deploy` method on a flow.

    ```python hl_lines="17-22" title="buy.py"
    from prefect import flow

    @flow(log_prints=True)
    def buy():
        print("Buying securities")

    if __name__ == "__main__":
        buy.deploy(
            name="my-code-baked-into-an-image-deployment", 
            work_pool_name="my-docker-pool", 
            image="my_registry/my_image:my_image_tag"
        )
    ```

>     Make sure you have the [work pool](/concepts/work-pools/) created in the Prefect Cloud workspace your are authenticated to or on your running self-hosted server instance connected to.  

Reviewer 2: ```suggestion
    Make sure you have the [work pool](/concepts/work-pools/) created in the Prefect Cloud workspace you are authenticated to or on your running self-hosted server instance.  
```

    You can specify a registry other than Docker Hub by providing the full registry path in the `image` argument.

    !!! warning
        If building a Docker image, the environment in which you are creating the deployment needs to have Docker installed and running.
        The [`prefect-docker` package](https://prefecthq.github.io/prefect-docker/) also needs to be installed.

    To avoid pushing to a registry, set `push=False` in the `.deploy` method.
    
    ```python hl_lines="6"

    if __name__ == "__main__":
        buy.deploy(
            name="my-code-baked-into-an-image-deployment", 
            work_pool_name="my-docker-pool", 
            image="my_registry/my_image:my_image_tag",
            push=False
        )
    ```

    To avoid building an image, set `build=False` in the `.deploy` method.

    ```python hl_lines="6"

    if __name__ == "__main__":
        buy.deploy(
            name="my-code-baked-into-an-image-deployment", 
            work_pool_name="my-docker-pool", 
            image="discdiver/no-build-image:1.0",
            build=False
        )
    ```

>     At runtime, the specified image will need to be available to access your flow code.

Reviewer 2: ```suggestion
    The specified image will need to be available in your deployment's execution environment for your flow code to be accessible.
```
Trying to differentiate that "runtime" is where the flow runs, not where you're running `.deploy`.

    See all the optional keyword arguments for the DeploymentImage class [here](https://docker-py.readthedocs.io/en/stable/images.html#docker.models.images.ImageCollection.build). 

    While baking code into Docker images is a popular deployment option, many teams decide to store their workflow code in git-based storage, such as GitHub, Bitbucket, or Gitlab. Let's see how to do that next.

    ### Store you code in git-based cloud storage 

    If you don't specify an `image` argument for `.deploy`, then you need to specify where to pull the flow code from at runtime with the `from_source` method. 

    Here's how we can pull our flow code from a GitHub repository.

    ```python hl_lines="4-6" title="git_storage.py" 
    from prefect import flow

    if __name__ == "__main__":
        flow.from_source(
            "https://github.com/my_github_account/my_repo/my_file.git",
            entrypoint="flows/no-image.py:hello_world",
        ).deploy(
            name="no-image-deployment",
            work_pool_name="my_pool",
            build=False
        )
    ```

>     The `entrypoint`` is the path to the file the flow is located in and the function name, separated by a colon.

Reviewer 2: ```suggestion
    The `entrypoint` is the path to the file the flow is located in and the function name, separated by a colon.
```

    - `pip_install_requirements` installs dependencies from a `requirements.txt` file within a specified directory.

    Below is an example of installing dependencies from a `requirements.txt` file after cloning:

    ```yaml
    pull:
        - prefect.deployments.steps.git_clone:
            id: clone-step  # needed in order to be referenced in subsequent steps
            repository: https://github.com/org/repo.git
        - prefect.deployments.steps.pip_install_requirements:
            directory: {{ clone-step.directory }}  # `clone-step` is a user-provided `id` field
            requirements_file: requirements.txt
    ```

    Below is an example that retrieves an access token from a 3rd party Key Vault and uses it in a private clone step:

    ```yaml
    pull:
    - prefect.deployments.steps.run_shell_script:
        id: get-access-token
        script: az keyvault secret show --name <secret name> --vault-name <secret vault> --query "value" --output tsv
        stream_output: false
    - prefect.deployments.steps.git_clone:
        repository: https://bitbucket.org/samples/deployments.git
        branch: master
        access_token: "{{ get-access-token.stdout }}"

>     ```

Reviewer 3: This is a great example!
